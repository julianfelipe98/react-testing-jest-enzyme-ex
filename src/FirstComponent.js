import React from "react";
import PropTypes from 'prop-types'
/* -------------------------------------------------------------------------- */
/*                            FUNCTIONAL COMPONENTE                           */
/* -------------------------------------------------------------------------- */
// no es comun usar props en los parametros del componente , hoy en dia lo que se hace es directamente una destructuracion de los datos que se mandan en las props
// para validar a que se tiene que enviar cierto parametro en las props, hay varias formas
const FirstComponent = ({ greeting,subtitle }) => {

  // dependiendo la manera en la que vayamos a usar el retorno hay que tener en cuenta que js en el return solo puede devolver un objeto , por lo cual si vamos a poner mas de una etiqueta en el return pues tendriamos que poner todo en un solo div , con esto suele generar el problema que podriamos llenarnos de muchos divs en el html y esto muchas veces no es lo deseado , para eso esta el fragmente que es una etiqueta propia de react que me permite retornar mas de una etiqueta pero que no me va a generar mas div en el cofigo , dependiendo del uso sirve , si el componente que se retorna necesita ciertas clases de css pues ahi si se podria meter en un div
  // react tiene una forma corta de importar fragment
  // y es solo abriendo y cerrando llaves sin ninguna etiqueta dentro de ellas
  // dentro de las llaves se puede escribir cualquier sentencia js que retorne un valor ( se puede imprimir muchas cosas pero no todo) por ejemplo los booleanos no los va a imprimir,los objetos tampoco se pueden imprimir de manera directa , se podria imprimir con un json.stringify ,
  return (
    <>
      <h1>{greeting}</h1>
      <p>{subtitle}</p>
    </>
  );
};
// de esta manera podemos indicar a otros usuarios u otros programadores de que tienen que mandar algo de cierta manera en las props
FirstComponent.propTypes={
    greeting:PropTypes.string.isRequired,
}
// de esta manera definimos un valor por defecto de las propiedades , en dado caso de que asi lo quisieramos y no quisieramos usar la forma que nos da js
// de igual manera otro manejo que se le suele dar a los valores por defecto es sencillamente dejarle un string vacio que al menos nos dejara ver en las herramientas de desarrollo de chrome que la prop era necesaria
FirstComponent.defaultProps={
    subtitle:'subtitle default  '
}
export default FirstComponent;
