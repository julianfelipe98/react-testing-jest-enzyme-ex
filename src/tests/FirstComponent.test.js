import { shallow } from "enzyme";
import FirstComponent from "../FirstComponent";
// realizo esta importacion dado a que si no lo importo no me reconoce la funcion de tobe in the document

describe("First component tests", () => {
  // test('Must to show the message "Hello World!"', () => {
  // const greeting = "Hello World!";
  // // el wrapper representa el producto renderizado de lo que yo quiero mostrar
  // const { getByText } = render(<FirstComponent greeting={greeting} />);
  // expect( getByText(greeting) ).toBeInTheDocument(true);

  // });
  test("it must to show <FirstComponent/> in the correct way", () => {
    // debemos asegurarnos de que se renderize con los valores que nosotros esperamos
    // shallow es muy parecido al render pero nos dara nuevas opciones como simular un click o tener acceso al documento si quisieramos trabajar como algun tipo de selector o algo asi
    const greeting = "hello world";
    const wrapper = shallow(<FirstComponent greeting={greeting} />);
    // hay que tener cuidado con la parte de las fotografias dado a que si cambiamos asi sea un espacio en el documento y queremos que asi sea la correcta pues tendriamos que actualizar el snapshot, este error que no arroja cada que actualizamos algo tambien nos permite tener una nocion de las diferencias que hay entre la fotografia que tenemos y el documento actual renderizado
    expect(wrapper).toMatchSnapshot();
  });
  test("must to show the sended subtitile by props", () => {
    const greeting = "hello world";
    const subtitle = "subtitle test";
    // hay que tener en cuenta que wrapper tiene toda la info del componente renderizado
    const wrapper = shallow(
      <FirstComponent greeting={greeting} subtitle={subtitle} />
    );
    // esta sentencia es muy similar al document.querySelector de js en el dom de un html normal
    // para este caso va a buscar todas las etiquetas p , como solo hay una y es la del subtitulo pues nos retorna esa , pero si hubiera mas nos retornaria un arreglo
    const textParagraf = wrapper.find("p").text();
    expect(textParagraf).toBe(subtitle);
  });
});
