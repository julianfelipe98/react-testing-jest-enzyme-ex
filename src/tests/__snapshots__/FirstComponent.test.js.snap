// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`First component tests it must to show <FirstComponent/> in the correct way 1`] = `
<Fragment>
  <h1>
    hello world
  </h1>
  <p>
    subtitle default  
  </p>
</Fragment>
`;
